8306461: ObjectInputStream::readObject() should handle negative array sizes without throwing NegativeArraySizeExceptions

simonis opened the pull request:
This issue was reported by: Yakov Shafranovich ([yakovsh@amazon.com](mailto:yakovsh@amazon.com))

Currently, `ObjectInputStream::readObject()` doesn't explicitly checks for a negative array length in the deserialization stream. Instead it calls `j.l.r.Array::newInstance(..)` with the negative length which results in a `NegativeArraySizeException`. NegativeArraySizeException is an unchecked exception which is neither declared in the signature of `ObjectInputStream::readObject()` nor mentioned in its API specification. It is therefore not obvious for users of `ObjectInputStream::readObject()` that they may have to handle `NegativeArraySizeException`s. It would therefor be better if a negative array length in the deserialization stream would be automatically wrapped in an `InvalidClassException` which is a checked exception (derived from `IOException` via `ObjectStreamException`) and declared in the signature of `ObjectInputStream::readObject()`.

If we do the negative array length check in `ObjectInputStream::readObject()` before filtering, this will then also fix `ObjectInputFilter.FilterInfo::arrayLength()` which is defined as:
```
Returns:
the non-negative number of array elements when deserializing an array of the class, otherwise -1
```
but currently returns a negative value if the array length is negative.

simonis commented:
/contributor add Yakov Shafranovich <yakovsh@amazon.com>

jddarcy commented:
Please file a CSR for the proposed behavioral change.

jddarcy commented:
/csr needed

simonis commented:
Hi @jddarcy,

I'm happy to create a CSR for this change, but I'm a little bit unsure about the details. From my understanding this qualifies as a behavioral change, right? But this behavior wasn't specified before at all. Neither did the API specification of `ObjectInputStream::readObject()` mention that it can throw a `NegativeArraySizeException` nor did the [Serialization Specification](https://docs.oracle.com/en/java/javase/20/docs/specs/serialization/index.html) mentioned the case of a negative array size.

Previously, `readObject()` could throw a `NegativeArraySizeException` which will now, with this PR, be changed into a `InvalidClassException`. Do you agree that this is an implementation detail and therefor the CSR should have "Implementation" Scope?

Thanks,
Volker

turbanoff commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,112 @@
+        serializedData[firstPos+4] = (byte) 0xff;
+        serializedData[firstPos+5] = (byte) 0xfe;
+
+        return  serializedData;
```
```suggestion
        return serializedData;
```

simonis replied:
Thanks, fixed.

shipilev commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,112 @@
+
+    public static void main(String[] args) throws Exception {
+        try {
+            test_simpleArray_negative();
```
Inline these `test_*` methods. Move the `CustomFilter` as test class nested.

simonis replied:
Done.

shipilev commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,112 @@
+        } catch (NegativeArraySizeException nase) {
+            throw new Exception("ObjectInputStream::readObject() shouldn't throw a NegativeArraySizeException", nase);
+        } catch (ObjectStreamException ose) {
+            // OK, because a NegativeArraySizeException should be converted into a ObjectStreamException
```
Check the message from `ose` here?

simonis replied:
Done. I only hope nobody will localize the exception message :)

shipilev commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,112 @@
+        } catch (NegativeArraySizeException nase) {
+            throw new Exception("ObjectInputStream::readObject() shouldn't throw a NegativeArraySizeException", nase);
+        } catch (ObjectStreamException ose) {
+            if (ose instanceof InvalidClassException ice && ice.getMessage().contains("filter status: REJECTED")) {
+                throw new Exception("ObjectInputStream::readObject() should catch NegativeArraySizeExceptions before filtering", ice);
+            }
+            // OK, because a NegativeArraySizeException should be converted into a ObjectStreamException *before* filtering
```
Sounds like we should really test for ICE with "Array length ... is negative" here. This would cover the case when filter rejected the class before that exception is thrown.

simonis replied:
The test for the `InvalidClassException` created by the object input filter is just above the comment, but I added another check for the correct exception message just for the simple test without filter.

shipilev reviewed (commented):
Some suggestions

jddarcy commented:
> Hi @jddarcy,
> 
> I'm happy to create a CSR for this change, but I'm a little bit unsure about the details. From my understanding this qualifies as a behavioral change, right? But this behavior wasn't specified before at all. Neither did the API specification of `ObjectInputStream::readObject()` mention that it can throw a `NegativeArraySizeException` nor did the [Serialization Specification](https://docs.oracle.com/en/java/javase/20/docs/specs/serialization/index.html) mentioned the case of a negative array size.
> 
> Previously, `readObject()` could throw a `NegativeArraySizeException` which will now, with this PR, be changed into a `InvalidClassException`. Do you agree that this is an implementation detail and therefor the CSR should have "Implementation" Scope?
> 
> Thanks, Volker

Hi Volker,

Yes, the behavior wasn't specified, but that doesn't imply users haven't become reliant on it, hence the (behavioral) compatibility review via a CSR of the implementation change.

HTH

RogerRiggs commented on src/java.base/share/classes/java/io/ObjectInputStream.java:
```diff
@@ -2138,7 +2138,9 @@ private Object readArray(boolean unshared) throws IOException {
         int len = bin.readInt();
-
+        if (len < 0) {
+            throw new InvalidClassException(desc.getName(), "Array length < 0 (" + len + ")");
```
I'd keep the exception logic simple/minimal.  Especially since this is a manual and exceptional case that is not the result of any input provided through an API.
```
throw new InvalidClassException(desc.getName(), "Array length is negative");
```

simonis replied:
Done.

RogerRiggs commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,112 @@
+            // 0x78 = TC_ENDBLOCKDATA
+            if (serializedData[i] == 0x78) {
+                firstPos = i;
+                break;
```
Move setting the length code (lines 64-67) here and return immediately.

simonis replied:
Fixed.

RogerRiggs commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,112 @@
+                break;
+            }
+        }
+
```
The test should fail if it falls through without finding the BLOCKDATA. Just a double guard against the unexpected.

simonis replied:
Fixed.

simonis commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,112 @@
+
+public class NegativeArraySizeTest {
+
+    private static byte[] _buildPayload() throws IOException {
```
Done.

simonis commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,112 @@
+public class NegativeArraySizeTest {
+
+    private static byte[] _buildPayload() throws IOException {
+        String[] simpleArray = new String[1];
```
Done.

simonis commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,112 @@
+
+        // Find the right location to modify, looking for the first instance of TC_ENDBLOCKDATA
+        int firstPos = 0;
+        for (int i=0; i<serializedData.length-1; i++) {
```
Done.

simonis commented:
Hi @turbanoff, @shipilev and @RogerRiggs,

Thanks for your reviews so far. I've hopefully addressed them all. I've also created a CSR for the issue:

https://bugs.openjdk.org/browse/JDK-8306744

Please feel  free to review it :)

RogerRiggs commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,109 @@
+            throw new Exception("ObjectInputStream::readObject() shouldn't throw a NegativeArraySizeException", nase);
+        } catch (ObjectStreamException ose) {
+            // OK, because a NegativeArraySizeException should be converted into a ObjectStreamException
+            if ("Array length is negative".equals(ose.getMessage())) {
```
The complete `getMessage()` is `[Ljava.lang.String;; Array length is negative`.
Perhaps: 
```
ose.getMessage().contains("Array length is negative") ...
```

simonis replied:
Your right. @RogerRiggs  I've further simplified the exception message to only contain the message "Array length is negative". I hope that's fine for you and in line with your previous comment to "*keep the exception logic simple/minimal*".

shipilev reviewed (approved):
Looks fine to me.

simonis commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,109 @@
+                throw new Exception("ObjectInputStream::readObject() should catch NegativeArraySizeExceptions before filtering", ice);
+            }
+            // OK, because a NegativeArraySizeException should be converted into a ObjectStreamException *before* filtering
+            if ("Array length is negative".equals(ose.getMessage())) {
```
Obviously!

Good catch. The problem was that the exception message also contained the name of the array class that's why *it worked*. I've now simplified to only contain the message "Array length is negative" which I think is in line with Roger'S previous comment to "*keep the exception logic simple/minimal*".

RogerRiggs reviewed (changes requested):
A bit more investigation is needed.   
I noticed that ObjectInputStream.checkArray throws NegativeArraySizeException if length is < 0 before calling filterCheck.
The checkArray method is called via SharedSecrets to check array sizes on some of the collection classes during deserialization.
Likely, the change should be extended to cover negative lengths in those cases too.

simonis commented:
Hi @RogerRiggs ,

I've checked the callers of `checkArray()` and found 13, all of them in the corresponding `readObject()` methods:

| class | NASE | exception |
|:----- |:----:|:--------- |
| `j.u.IdentityHashMap` | + | StreamCorruptedException("Illegal mappings count: " + size) |
| `j.u.HashMap` | + | InvalidObjectException("Illegal mappings count: " + mappings) |
| `j.u.Properties` | + | StreamCorruptedException("Illegal # of Elements: " + elements) |
| `j.u.ArrayList` | + | InvalidObjectException("Invalid size: " + size) |
| `j.u.ImmutableCollections` | + | InvalidObjectException("negative length " + len) |
| `j.u.Hashtable` | + | StreamCorruptedException("Illegal # of Elements: " + elements) |
| `j.u.HashSet` | + | InvalidObjectException("Illegal capacity: " + capacity) |
| `jx.m.o.TabularDataSupport` | (+) | (depends on ArrayList) |
| `j.u.concurrent/PriorityBlockingQueue` | (-) | (depends on PriorityQueue) |
| `j.u.PriorityQueue` | - | |
| `j.u.ArrayDeque` | - | |
| `j.u.concurrent/CopyOnWriteArrayList` | - | |
| `j.u.Collections` | - | |

The occurrences with a "+" int he "NASE" column all already handle the negative array size case **before** calling `ObjectInputStream.checkArray()` and throw the exceptions listed in the "exception" column if the array size is negative.

So instead of changing `ObjectInputStream.checkArray()` maybe it makes sense to fix the remaining four classes (i.e. `PriorityQueue`, `ArrayDeque`, `CopyOnWriteArrayList` and `Collections`) to handle negative array sizes in their corresponding  `readObject()` methods as well? This somehow feels more natural to me, because, other then within the initial issue, this is not a problem of the deserialization protocol, but rather  an issue with custom deserialization code in the the `readObject()` methods. Therefor I also think it would make sense to handle that as a different issue. What do you think?

RogerRiggs commented:
Thanks for the investigation.

On the question of the exception thrown IllegalObjectException vs StreamCorruptedExeception, I'd lean toward StreamCorruptedException, including for the current PR; as that is more indicative of the issue raised.

As for addressing the existing uses of checkArray that would throw NAE, I would rather see a single fix in checkArray than adding code in multiple other places.  A fix in checkArray would cover future uses as well as current uses. The existing code that is checking len < 0 before calling checkArray can continue to do so to maintain compatibility on the exception thrown. Though a change would be unlikely to break user code it is better to avoid that. (It might cause changes existing tests).

Handling it in a separate PR may be reasonable but it too will require a CSR (change in behavior from throwning NAE to SCE) and the cause and behavior change are generally the same as the current PR.  If handled in a single PR/CSR and release note will have a more coherent and singular explanation.

simonis commented:
Hi @RogerRiggs,

I have now updated both, `checkArray()` and `readArray()` to throw a `StreamCorruptedException("Array length is negative")` in the case of a negative array size. I've also changed the signature of `ObjectInputStream::checkArray()` and `JavaObjectInputStreamAccess::checkArray` to throw an `ObjectStreamException` instead of an `InvalidClassException` because `ObjectStreamException` is the super class of both `InvalidClassException` and `StreamCorruptedException`. Finally I've updated the test with an extra case of a `PriorityQueue` with a negative element size in order to test `checkArray()`.

OK now? Once we read consensus here I'll update the CSR accordingly.

shipilev commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,147 @@
+
+public class NegativeArraySizeTest {
+
+    private static byte[] buildArrayPayload() throws IOException {
+         // Serialize to bytes
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
```suggestion
    private static byte[] buildArrayPayload() throws IOException {
        // Serialize to bytes
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
```

simonis replied:
You're right. An for some strange reason the same comment  in `buildPriorityQueuePayload()` is correctly indented but the following code was not. So fixed that as well.

RogerRiggs reviewed (approved):
Look good, thanks for fixing checkArray().

shipilev commented on test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    public static void main(String[] args) throws Exception {
+        try {
```
Test code nits, we can probably `try-with-resources` this thing to make it cleaner and make sure nothing unexpectedly bad is happening on "close" paths.

simonis replied:
OK, here we go...

shipilev reviewed (approved):
`StreamCorruptedException` looks better. I have a few nitpicks.

simonis commented:
I've created [JDK-8307621](https://bugs.openjdk.org/browse/JDK-8307621) to track the release note for this issue.

simonis commented:
/integrate
